### PR TITLE
BWA-124 - 'Copy' Option Missing from Long-Press Menu

### PR DESCRIPTION
**Pull request by phil-livefront**
## 🎟️ Tracking

[BWA-124](https://bitwarden.atlassian.net/browse/BWA-124)

## 📔 Objective

- Implemented the option to copy an item from the long-press menu. While items could already be copied by clicking on them, adding this functionality to the menu was straightforward.
- Renamed "Copy to Bitwarden" to "Move to Bitwarden" to align with the updated Figma design.
- Refactored the action handling logic to use an enum for dropdown menu actions, consolidating logic and improving maintainability. Added missing tests to cover the new functionality.

## 📸 Screenshots

https://github.com/user-attachments/assets/9f680a5e-3faf-4acf-bbf1-65663bdf1f86

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[BWA-124]: https://bitwarden.atlassian.net/browse/BWA-124?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 1]
```diff
@@ -142,6 +142,12 @@ class ItemListingViewModel @Inject constructor(
                 handleInternalAction(action)
             }
 
+            is ItemListingAction.DropdownMenuClick -> {
+                viewModelScope.launch {
+                    handleDropdownMenuClick(action)
```
[the remark this answers is not in the document]
Comment (phil-livefront) [reply app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:147]: ah i see!  will do, thanks!

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 2]
```diff
@@ -528,6 +534,17 @@ class ItemListingViewModel @Inject constructor(
         sendEvent(ItemListingEvent.NavigateToBitwardenListing)
     }
 
+    private suspend fun handleDropdownMenuClick(action: ItemListingAction.DropdownMenuClick) {
+        sendAction(
+            when (action.menuAction) {
+                VaultDropdownMenuAction.COPY -> ItemListingAction.ItemClick(action.id)
```
Comment (david-livefront) [inline app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:540]: We do not need to send actions like this, just call the functions (no need to suspend).

I am also wondering if wee need the actions at all?

File: app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingScreenTest.kt [hunk 3]
```diff
@@ -222,7 +222,9 @@ class ItemListingScreenTest : BaseComposeTest() {
 
         verify {
             viewModel.trySendAction(
-                ItemListingAction.DropdownMenuClick(VaultDropdownMenuAction.MOVE, "1"),
+                ItemListingAction.DropdownMenuClick(
+                    VaultDropdownMenuAction.MOVE, LOCAL_CODE
```
Comment (david-livefront) [inline app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingScreenTest.kt]: Formatting:

```kotlin
ItemListingAction.DropdownMenuClick(
    VaultDropdownMenuAction.MOVE,
    LOCAL_CODE,
)
```

File: app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt [hunk 4]
```diff
@@ -469,17 +473,16 @@ class ItemListingViewModelTest : BaseViewModelTest() {
     @Test
     fun `should trigger delete prompt when DropdownMenuClick DELETE is triggered`() = runTest {
         val viewModel = createViewModel()
-        val testId = "123456"
 
         val expectedState = DEFAULT_STATE.copy(
             dialog = ItemListingState.DialogState.DeleteConfirmationPrompt(
                 message = R.string.do_you_really_want_to_permanently_delete_cipher.asText(),
-                testId,
+                LOCAL_CODE.id,
```
Comment (david-livefront) [inline app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt:480]: Can we add a name to this param

File: app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt [hunk 5]
```diff
@@ -382,7 +383,9 @@ class ItemListingViewModelTest : BaseViewModelTest() {
 
         val viewModel = createViewModel()
 
-        viewModel.trySendAction(ItemListingAction.MoveToBitwardenClick(entityId = "1"))
+        viewModel.trySendAction(
+            ItemListingAction.DropdownMenuClick(VaultDropdownMenuAction.MOVE, item = LOCAL_CODE)
```
Comment (david-livefront) [inline app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt]: These should both have named params

**Review by david-livefront — APPROVED** [review]

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 6]
```diff
@@ -123,11 +119,7 @@ class ItemListingViewModel @Inject constructor(
             }
 
             is ItemListingAction.ItemClick -> {
-                handleItemClick(action)
-            }
-
-            is ItemListingAction.EditItemClick -> {
-                handleEditItemClick(action)
+                handleItemClick(action.authCode)
```
Comment (dseverns-livefront) [inline app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:122]: Not a blocker for me, but more a discussion. @david-livefront on the password manager I know we have been avoiding unwrapping actions at this level and always passing the full action into the `handle*()` function. I am seeing there are somethings in the `when` that already kind of break away from some of those conventions like sending events, but curious if we want to keep that pattern up in Auth too.

File: app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt [hunk 7]
```diff
@@ -423,6 +432,68 @@ class ItemListViewModelTest : BaseViewModelTest() {
...
+
+        val expectedState = DEFAULT_STATE.copy(
+            dialog = ItemListingState.DialogState.DeleteConfirmationPrompt(
+                message = R.string.do_you_really_want_to_permanently_delete_cipher.asText(),
+                itemId = LOCAL_CODE.id,
+            ),
+        )
+
+        viewModel.trySendAction(
+            ItemListingAction.DropdownMenuClick(VaultDropdownMenuAction.DELETE, LOCAL_CODE),
```
Comment (dseverns-livefront) [inline app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt:488]: ⛏️ named args

File: app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt [hunk 8]
```diff
@@ -423,6 +432,68 @@ class ItemListViewModelTest : BaseViewModelTest() {
...
+    @Test
+    fun `should copy text to clipboard when DropdownMenuClick COPY is triggered`() = runTest {
+        val viewModel = createViewModel()
+        val testId = "123456"
+
+        every { clipboardManager.setText(text = testId) } just runs
+
+        viewModel.eventFlow.test {
+            viewModel.trySendAction(
+                ItemListingAction.DropdownMenuClick(VaultDropdownMenuAction.COPY, LOCAL_CODE),
```
Comment (dseverns-livefront) [inline app/src/test/java/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModelTest.kt:444]: ⛏️ named args here too :)

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt (hunk 6 above)
Comment (david-livefront) [reply app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:122]: That is a good point, it is a little odd here that these are not actions though. I wonder if the `handle` moniker is the problem 🤔

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 9]
```diff
@@ -157,38 +153,34 @@ class ItemListingViewModel @Inject constructor(
...
+    private fun handleMoveToBitwardenClick(itemId: String) {
         viewModelScope.launch {
             val item = authenticatorRepository
-                .getItemStateFlow(action.entityId)
+                .getItemStateFlow(itemId)
                 .first { it.data != null }
 
             val didLaunchAddTotpFlow = authenticatorBridgeManager.startAddTotpLoginItemFlow(
-                totpUri = item.data!!.toOtpAuthUriString(),
+                totpUri = item.data?.toOtpAuthUriString() ?: "",
```
Comment (dseverns-livefront) [inline app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:183]: we can replace the `?: "" ` with `.orEmpty()`

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 10]
```diff
@@ -157,38 +153,29 @@ class ItemListingViewModel @Inject constructor(
...
     }
 
-    private fun handleItemClick(action: ItemListingAction.ItemClick) {
-        clipboardManager.setText(action.authCode)
-        sendEvent(
-            ItemListingEvent.ShowToast(
-                message = R.string.value_has_been_copied.asText(action.authCode),
-            ),
-        )
+    private fun handleItemClick(authCode: String) {
```
Comment (david-livefront) [inline app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:163]: Can we add `copy` to this function name

Comment (phil-livefront) [reply app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:163]: yeah good call

File: app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt [hunk 11]
```diff
@@ -892,15 +878,25 @@ sealed class ItemListingAction {
...
+    data class ConfirmDeleteClick(val itemId: String) : ItemListingAction()
+
+    /**
+     * Represents an action triggered when the user clicks an item in the dropdown menu.
+     *
+     * @param menuAction The action selected from the dropdown menu.
+     * @param id The identifier of the item on which the action is being performed.
      */
-    data class MoveToBitwardenClick(val entityId: String) : ItemListingAction()
+    data class DropdownMenuClick(
```
Comment (SaintPatrck) [inline app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:891]: ❤️

**Review by SaintPatrck — APPROVED** [review]

**Review by dseverns-livefront — APPROVED** [review]

**Pull request merged**